tests/C0/UsernameValidationTest.php

```diff
@@ -2,9 +2,67 @@
 
 namespace Tests\C0;
 
+use App\UsernameValidation;
 use PHPUnit\Framework\TestCase;
 
 class UsernameValidationTest extends TestCase
 {
-    // TODO: write 6 tests to coverage 100% code
+    function generateRandomString($length = 10) {
+        return substr(str_shuffle(str_repeat($x = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ', ceil($length/strlen($x)))), 1, $length);
+    }
+
+    public function test_give_minimum_username_then_false_and_show_message()
+    {
+        $usernameValidation = new UsernameValidation();
+
+        $this->assertFalse($usernameValidation->isValid(''));
+        $this->assertEquals('Minimum length is ' . UsernameValidation::MIN_LENGTH, $usernameValidation->getMessage());
+    }
+
+    public function test_give_maximum_username_then_false_and_show_message()
+    {
+        $usernameValidation = new UsernameValidation();
+        $username = $this->generateRandomString(UsernameValidation::MAX_LENGTH + 1); //get string always greater then MAX_LENGTH
+
+        $this->assertFalse($usernameValidation->isValid($username));
+        $this->assertEquals('Maximum length is ' . UsernameValidation::MAX_LENGTH, $usernameValidation->getMessage());
+    }
+
+    public function test_username_have_dash_at_begin_or_end_then_false_and_show_message()
```

Comment:
So với cách đặt tên:
```
test_IT_not_pass_WHEN_value_has_dash_at_begining
```
Thì có dễ hiểu hơn không nhỉ? Ở đây IT là class under test => `it not pass` nghĩa là `usernam validation not pass` WHEN là điều kiện đầu vào

Khi dùng `--test-dox` ta sẽ có message: 
```
Username Validation (Tests\C0\UsernameValidation)
 ✔ IT not pass WHEN value exceed maxlength
 ✔ IT not pass WHEN value has dash at begining
 ✔ IT not pass WHEN value has multiple dashes
```

Reply:
@tuanpt-0634 
VIết như anh cũng được, nói chung tùy vào ngữ cảnh. Nhưng trên thực tế thì sử dụng từ `pass` hoặc từ `it` thì nó khá là ít ngữ cảnh để dùng và thực ra nó khá là chung chung. Cái ở đây em muốn mọi người nhìn vào có thể hiểu ngay lập tức function này test gì chứ ko phải là nhìn vào function xong nhìn thêm vào class nữa vì trong trường hợp 1 class nhiều case thì giả sử nhé số case xuống dòng vượt quá chiều cao của màn hình => anh phải khéo lên tìm cái class của nó để xem nó đang test cái gì...

Reply:
`pass` ở đây dùng cho class validation thôi, vì validation thì chỉ có pass hay failed. Còn anh đang muốn suggest chung 1 pattern đặt tên cho test, ko bắt buộc nhưng dễ follow lúc đầu. Ở đây anh nghĩ nên xem class là `unit` thay vì từng method, khi đó mình sẽ test hành vi của class thay vì test từng method, chẳng hạn có thay đổi tên method nhưng có thể không cần đặt lại tên test case, anh đúc kết từ bài này https://dzone.com/articles/7-popular-unit-test-naming

```php
Should_ExpectedBehavior_When_StateUnderTest() {}
Should_ThrowException_When_AgeLessThan18() {}
Should_FailToWithdrawMoney_ForInvalidAccount() {}
Should_FailToAdmit_IfMandatoryFieldsAreMissing() {}
```
=>
```php
public function test_it_expected_behavior_when_condition() {}
```

Reply:
oki anh, thế thì mình nên tổng hợp lại rồi để đó là 1 convention ạ . Kiểu thế thì mọi người cũng dễ follow hơn

Reply:
anw nếu mọi người quen rồi thì đặt như thế nào cũng được mà a :))), Freestyle mà :v: